### ISSUE TITLE
[Bug] Passive mechanic breaks after a canceled Death trigger

### ISSUE THREAD

**Issue by Flare2B** [promcteam/proskillapi#44]
**Describe the bug**
Cancelling a Death trigger with the Cancel mechanic will break Passive tasks running on the player.

**To Reproduce**
![immagine](https://user-images.githubusercontent.com/30595290/126365923-6a16c3df-a32d-4c1d-9f45-22249db91d79.png)

**Expected behavior**
Passive tasks should still run after a canceled Death trigger

**latest.log**
No errors in the log, clean server install with no other plugins installed.

**Versions (please complete the following information):**
-  Spigot Version: Paper #699
 - Server Version: 1.16.4
 - Plugin Version: Alpha 1.1.6

**Additional info**
Most likely an issue in the PassiveTask class (defined in src/main/java/com/sucy/skill/dynamic/mechanic/PassiveMechanic.java), which assumes the caster (player) is dead, despite the death event being cancelled.

**Comment by goflishMC** [comment]
> **Describe the bug**
> Cancelling a Death trigger with the Cancel mechanic will break Passive tasks running on the player.
> 
> **To Reproduce**
> ![immagine](https://user-images.githubusercontent.com/30595290/126365923-6a16c3df-a32d-4c1d-9f45-22249db91d79.png)
> 
> **Expected behavior**
> Passive tasks should still run after a canceled Death trigger
> 
> **latest.log**
> No errors in the log, clean server install with no other plugins installed.
> 
> **Versions (please complete the following information):**
> 
> * Spigot Version: Paper #699
> * Server Version: 1.16.4
> * Plugin Version: Alpha 1.1.6
> 
> **Additional info**
> Most likely an issue in the PassiveTask class (defined in src/main/java/com/sucy/skill/dynamic/mechanic/PassiveMechanic.java), which assumes the caster (player) is dead, despite the death event being cancelled.

if you can send me your skill file I will test this.

**Comment by Flare2B** [comment]
> 
> 
> > **Describe the bug**
> > Cancelling a Death trigger with the Cancel mechanic will break Passive tasks running on the player.
> > **To Reproduce**
> > ![immagine](https://user-images.githubusercontent.com/30595290/126365923-6a16c3df-a32d-4c1d-9f45-22249db91d79.png)
> > **Expected behavior**
> > Passive tasks should still run after a canceled Death trigger
> > **latest.log**
> > No errors in the log, clean server install with no other plugins installed.
> > **Versions (please complete the following information):**
> > 
> > * Spigot Version: Paper #699
> > * Server Version: 1.16.4
> > * Plugin Version: Alpha 1.1.6
> > 
> > **Additional info**
> > Most likely an issue in the PassiveTask class (defined in src/main/java/com/sucy/skill/dynamic/mechanic/PassiveMechanic.java), which assumes the caster (player) is dead, despite the death event being cancelled.
> 
> if you can send me your skill file I will test this.

I'll send the file once I'm back from vacations, though this can be done with any passive task and a cancel death trigger as seen in the screenshot above.

**Comment by goflishMC** [comment]
Are you able to send me this skill now?

**Comment by Flare2B** [comment]
[Test-Bug.zip](https://github.com/promcteam/proskillapi/files/6991533/Test-Bug.zip)

Will heal the player for 25% health every 2 seconds. When the player dies and the Death trigger is cancelled, the heals stop working until /class reload or until the player actually dies.

**Comment by Travja** [comment]
I dug into this a little further and found a couple things lurking in the code. First and foremost, [PlayerDeathEvent](https://hub.spigotmc.org/javadocs/bukkit/org/bukkit/event/entity/PlayerDeathEvent.html) is not a cancellable event, therefore the cancel mechanic on this trigger would actually result in cancelling the skill itself, resulting in the behavior you're talking about.

A little bit of information I found while digging in the code... If the cancel mechanic is used on a trigger (or event) that is cancellable, the event will be cancelled and everything else should remain intact. For example, if I use a Block Break Trigger and a cancel mechanic, any passives running alongside should be fine as the block break will be cancelled (because it's cancellable). No on the other hand, since the Death Event is not cancellable, it will cancel the skill in its entirety.

If you're able to use the skill pictured above to cheat death, I'd be surprised since nothing I'm seeing at the moment in the code would suggest that you should be able to cancel the death in this way.

Now, a little bit of hope. It would be possible to create a sort of LethalBlow trigger that would be fired when a user receives a hit that would do more damage than they have health. The EntityDamageEvent is cancellable, so you could cheat death in this way by using the Cancel mechanic. If this is something you'd like to request, I'm going to close this ticket and you can create a separate request ticket for the feature. Just make sure to mention #44 when you do so we can reference anything we might need from here.

tldr: Unfortunately this is working as intended (and unfortunately can't really be 'fixed' as your implementation would require). But this is something we can create a workaround for.
